nmigen/lib/fifo.py

```diff
@@ -366,6 +367,31 @@ def elaborate(self, platform):
             self.r_rdy.eq(~r_empty),
         ]
 
+        # Reset handling to maintain FIFO and CDC invariants in the presence of a write-domain
+        # reset. For further discussion, see https://github.com/nmigen/nmigen/issues/181.
+        w_rst = ResetSignal(domain=self._w_domain, allow_reset_less=True)
+        r_rst = Signal()
+        
+        # Create synthetic clock domain to work around ResetSynchronizer expecting a domain
+        fake_domain = ClockDomain(name="rst_cdc", async_reset=True, local=True)
```

Comment:
I think this characterization isn't correct. You are adding some FFs with a different control set (one that includes w_rst), which in fact form a (very small) new clock domain. This doesn't inherently have anything to do with the way `ResetSynchronizer` works; nMigen requires you to create a clock domain for every distinct async control set.

Reply:
But `ResetSynchronizer` creates such a clock domain internally (right?), so this one is unnecessary.

Reply:
You're totally right. I was thinking of `ResetSynchronizer` delegating to platform code, in which case it generally won't use an nMigen clock domain, but that's a different problem.

Do you feel like fixing it?

Reply:
Happy to but the right solution wasn't obvious to me, other than making something like a `AsyncAssertSyncDeassertSynchronizer` module which `ResetSynchronizer` would use under the hood, and I didn't want to try to create a new cdc module without checking with you.

Reply:
What about (a) adding a `rst` attribute (output), alongside `arst` (input)`, and allowing (b) passing `domain=None` to the constructor? Then the platform code would assert `reset_sync.rst`, which in the case when domain is specified will be assigned to the domain.

Reply:
That should work. It seems a bit more complex to me, and it doesn't address a case where you might want to use this functionality for something besides a reset (admittedly I don't know what use case that would be), but it's a nice way to enhance `ResetSynchronizer` without (I think) having to touch all the platform code.

I lean towards creating the module I described (a more reasonable name might be `ReleaseSynchronizer`), but I'm open to the solution you described as well. I just think a generic module that gets replaced with platform-specific modules if available is a cleaner mental model.

Reply:
> without (I think) having to touch all the platform code.

You have to touch all the platform code with my suggestion.

> I just think a generic module that gets replaced with platform-specific modules if available is a cleaner mental model.

I'm a bit confused here. `ResetSynchronizer` *does* get replaced with platform-specific modules; that's half the reason for its existence and being in `nmigen.lib`.

If you mean that you want to make `ResetSynchronizer` a thin wrapper around some other module that just does `ResetSignal(self._domain).eq(...)` then I'm not really opposed to it, but it seems like the overhead (having two modules that do *almost* the same thing, duplicate constructor arguments, documentation, etc) is not worth the benefit (maybe slightly cleaner separation of concerns).

> admittedly I don't know what use case that would be

Yes. "Chain of FFs with asynchronous preset and synchronous clear" is a very specific use case. Also, sometime in the future we'll have to extend it to handle negative polarity resets (#185).

Reply:
Let me just push what I'm talking about and we can discuss it directly instead of each talking about the picture in our head of the picture in the other person's head :) can always roll back.